libkbfs/keybase_daemon_local.go

```diff
@@ -236,6 +250,146 @@ func (k *KeybaseDaemonLocal) Identify(
+// for KeybaseDaemonLocal.
+func (k *KeybaseDaemonLocal) ResolveIdentifyImplicitTeam(
+	ctx context.Context, assertions, suffix string, tlfType tlf.Type,
+	doIdentifies bool, reason string) (ImplicitTeamInfo, error) {
+	if err := checkContext(ctx); err != nil {
+		return ImplicitTeamInfo{}, err
+	}
+
+	if tlfType != tlf.Private && tlfType != tlf.Public {
+		return ImplicitTeamInfo{}, fmt.Errorf(
+			"Invalid implicit team TLF type: %s", tlfType)
+	}
+
+	k.lock.Lock()
+	defer k.lock.Unlock()
+
+	// Canonicalize the name.
+	writerNames, readerNames, _, err :=
+		splitAndNormalizeTLFName(assertions, tlfType)
+	if err != nil {
+		return ImplicitTeamInfo{}, err
+	}
+	var writers, readers []libkb.NormalizedUsername
+	var unresolvedWriters, unresolvedReaders []keybase1.SocialAssertion
+	resolvedIDs := make(map[libkb.NormalizedUsername]keybase1.UserOrTeamID)
+	for _, w := range writerNames {
+		writers, unresolvedWriters, err = k.resolveForImplicitTeam(
+			ctx, w, writers, unresolvedWriters, resolvedIDs)
+		if err != nil {
+			return ImplicitTeamInfo{}, err
+		}
+	}
+	for _, r := range readerNames {
+		readers, unresolvedReaders, err = k.resolveForImplicitTeam(
+			ctx, r, readers, unresolvedReaders, resolvedIDs)
+		if err != nil {
+			return ImplicitTeamInfo{}, err
+		}
+	}
+
+	var extensions []tlf.HandleExtension
+	if len(suffix) != 0 {
+		extensions, err = tlf.ParseHandleExtensionSuffix(suffix)
+		if err != nil {
+			return ImplicitTeamInfo{}, err
+		}
+	}
+	name := tlf.MakeCanonicalName(
+		writers, unresolvedWriters, readers, unresolvedReaders, extensions)
+
+	key := fmt.Sprintf("%s:%s", tlfType.String(), name)
+	tid, ok := k.implicitAsserts[key]
+	if ok {
+		return k.localImplicitTeams[tid], nil
+	}
+
+	// If the implicit team doesn't exist, always create it.
+
+	// Need to make the team info as well, so get the list of user
+	// names and resolve them.
+	asUserName := libkb.NormalizedUsername(name)
+	teams := makeLocalTeams(
+		[]libkb.NormalizedUsername{asUserName}, len(k.localTeams), tlfType)
+	info := teams[0]
+	info.Writers = make(map[keybase1.UID]bool, len(writerNames))
+	for _, w := range writers {
+		id, ok := resolvedIDs[w]
+		if !ok {
+			return ImplicitTeamInfo{}, fmt.Errorf("No resolved writer %s", w)
+		}
+		info.Writers[id.AsUserOrBust()] = true
+	}
+	if len(readerNames) > 0 {
+		info.Readers = make(map[keybase1.UID]bool, len(readerNames))
+		for _, r := range readers {
+			id, ok := resolvedIDs[r]
+			if !ok {
+				return ImplicitTeamInfo{}, fmt.Errorf(
+					"No resolved reader %s", r)
+
+			}
+			info.Readers[id.AsUserOrBust()] = true
+		}
+	}
+	// Unresolved users don't need to go in the team info, they're
+	// irrelvant until they're resolved.  TODO: add resolved users
+	// into existing teams they should be on.
+
+	tid = teams[0].TID
+	k.implicitAsserts[key] = tid
+	k.localTeams[tid] = info
+
+	iteamInfo := ImplicitTeamInfo{
+		// TODO: allow display name to differ?
```

Comment:
Could you elaborate a bit on this TODO? What would a differed name be and what is it for?

Reply:
There are two things here.  One is that implicit teams have an auto-generated name in `TeamInfo` that something like "_keybase_implicit_team_XXX" (I forget the format exactly).  It's only used internally and shouldn't be shown to users.    I should probably put a different TODO for that up where the `TeamInfo` is constructed.

The second thing is that the "display name" is the _preferred_ canonical name, with the logged-in user listed first, and all the other users in alphabetical order.  The `MakeCanonicalName` above doesn't capture this. I'll update the comment.

Reply:
I see. Thanks. How is the auto-generated name relevant here? This one is always display name right?

Reply:
Yep, I meant the "to differ" part in the comment has two meaning, one of which is that it differs from the implicit team name.  But as I said, I should (and did) fix that above where the team info is made.  I just had it create an implicit team name there.